File: Regularization/py/RidgeRegression.py
import numpy as np
def load_data(file):
	try:
		matrix = []
		f = open(file,'r')
		for line in f.readlines():
			x = list(map(float,line.split()))
			x.insert(0,1.0)
			matrix.append(x)
	finally:
		if f:
			f.close()
	return matrix

def sign(n):
	if n <=0:
		return -1
	else:
		return 1

def calErrRate(output,predict):
	errCount = 0
	for i in range(len(output)):
		if output[i]!=predict[i]:
			errCount = errCount + 1
	return errCount/len(output)

data = load_data('train.txt')
test = load_data('test.txt')
rowNum = len(data)
rowNum2 = len(test)
colNum = len(data[0])
colNum2 = len(test[0])
trainx =[]
trainy=[]
for i in data:
	trainx.append(i[:colNum-1])
	trainy.append(i[colNum-1])
testx =[]
testy=[]
for i in test:
	testx.append(i[:colNum-1])
	testy.append(i[colNum-1])
weight = np.dot(np.dot(np.linalg.inv(np.add(np.dot(np.transpose(trainx),trainx),np.dot(10,np.identity(colNum-1)))),np.transpose(trainx)),trainy)
trainpredict = list(map(sign,np.dot(weight,np.transpose(trainx))))
print("error rate in sample:",calErrRate(trainy,trainpredict))
testpredict = list(map(sign,np.dot(weight,np.transpose(testx))))
print("error rate out sample:",calErrRate(testy,testpredict))

File: Regularization/java/RidgeRegression/src/RidgeRegression.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

import org.ejml.data.DenseMatrix64F;
import org.ejml.simple.SimpleMatrix;


public class RidgeRegression {
	public static void main(String[] args){
		SimpleMatrix trainDataSimpleMatrix = loadData("train.txt");
		SimpleMatrix testDataSimpleMatrix = loadData("test.txt");
		SimpleMatrix weight = getRidgeRegressionWeight(trainDataSimpleMatrix);
		double errInRate = calculatErrorRate(weight,trainDataSimpleMatrix);
		double errOutRate = calculatErrorRate(weight,testDataSimpleMatrix);
		System.out.println("error rate in sample:"+errInRate);
		System.out.println("error rate out sample:"+errOutRate);
	}
	public static double sign(double n){
		if (n<=0)
			return -1;
		else
			return 1;
	}
	public static double calculatErrorRate(SimpleMatrix w,SimpleMatrix data){
		int rowNum = data.numRows();
		int colNum = data.numCols();
		SimpleMatrix xdata = data.extractMatrix(0, rowNum, 0, colNum-1);
		SimpleMatrix ydata = data.extractVector(false, colNum-1);
		int errCount = 0;
		for (int i=0;i<rowNum;i++){
			double predict = sign(xdata.extractVector(true, i).dot(w.transpose()));
			if (predict!=ydata.get(i))
				errCount++;
;		}
		return (double) errCount/rowNum;
	}
	
	public static SimpleMatrix getRidgeRegressionWeight(SimpleMatrix data){
		int rowNum = data.numRows();
		int colNum = data.numCols();
		SimpleMatrix xdata = data.extractMatrix(0, rowNum, 0, colNum-1);
		SimpleMatrix ydata = data.extractVector(false, colNum-1);
		SimpleMatrix weight = xdata.transpose().mult(xdata).plus(SimpleMatrix.identity(colNum-1).scale(10)).invert().mult(xdata.transpose()).mult(ydata);
		return weight;
	}
	public static SimpleMatrix loadData(String fileName){
		List<String> dataList = new ArrayList<String>();
		BufferedReader br = null;
		try {
				br = new BufferedReader(new FileReader(fileName));
				String line = null;
				while((line = br.readLine())!=null){
					dataList.add(line.trim());
				}
				br.close();
		} catch (IOException e){
			e.printStackTrace();
		}
		int rowNum = dataList.size();
		int colNum = dataList.get(0).split("\\s+").length+1;
		double[] dataMatrix = new double[rowNum*colNum];
		for (int i=0;i<rowNum;i++){
			dataMatrix[i*colNum] = 1;
			String[] newLine = dataList.get(i).split("\\s+");
			//System.out.println(dataList.get(i));
			for (int j=0;j<newLine.length;j++){
				dataMatrix[i*colNum+j+1] = Double.valueOf(newLine[j]);
			}
		}
		DenseMatrix64F dataDenseMatrix = new DenseMatrix64F(rowNum,colNum,true,dataMatrix);
		SimpleMatrix dataSimpleMatrix = SimpleMatrix.wrap(dataDenseMatrix);
		return dataSimpleMatrix;
	}
}
